plugin/storage/badger/spanstore/reader.go

```diff
@@ -0,0 +1,488 @@
+	ErrDurationMinGreaterThanMax = errors.New("Duration Minimum is above Maximum")
+
+	// ErrMalformedRequestObject occurs when a request object is nil
+	ErrMalformedRequestObject = errors.New("Malformed request object")
+
+	// ErrStartAndEndTimeNotSet occurs when start time and end time are not set
+	ErrStartAndEndTimeNotSet = errors.New("Start and End Time must be set")
+
+	// ErrUnableToFindTraceIDAggregation occurs when an aggregation query for TraceIDs fail.
+	ErrUnableToFindTraceIDAggregation = errors.New("Could not find aggregation of traceIDs")
+
+	// ErrNotSupported during development, don't support every option - yet
+	ErrNotSupported = errors.New("This query parameter is not supported yet")
+
+	errNoTraces = errors.New("No trace with that ID found")
+
+	defaultMaxDuration = model.DurationAsMicroseconds(time.Hour * 24)
+)
+
+const (
+	defaultNumTraces = 100
+	sizeOfTraceID    = 16
+)
+
+// TraceReader reads traces from the local badger store
+type TraceReader struct {
+	store *badger.DB
+	cache *CacheStore
+}
+
+// NewTraceReader returns a TraceReader with cache
+func NewTraceReader(db *badger.DB, c *CacheStore) *TraceReader {
+	return &TraceReader{
+		store: db,
+		cache: c,
+	}
+}
+
+// getTraces enriches TraceIDs to Traces
+func (r *TraceReader) getTraces(traceIDs []model.TraceID) ([]*model.Trace, error) {
+	// Get by PK
+	traces := make([]*model.Trace, 0, len(traceIDs))
+	prefixes := make([][]byte, 0, len(traceIDs))
+
+	for _, traceID := range traceIDs {
+		prefixes = append(prefixes, createPrimaryKeySeekPrefix(traceID))
+	}
+
+	err := r.store.View(func(txn *badger.Txn) error {
+		opts := badger.DefaultIteratorOptions
+		opts.PrefetchSize = 10 // TraceIDs are not sorted, pointless to prefetch large amount of values
+		it := txn.NewIterator(opts)
+		defer it.Close()
+
+		val := []byte{}
+		for _, prefix := range prefixes {
+			spans := make([]*model.Span, 0, 4) // reduce reallocation requirements by defining some initial length
+
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				// Add value to the span store (decode from JSON / defined encoding first)
+				// These are in the correct order because of the sorted nature
+				item := it.Item()
+				val, err := item.ValueCopy(val)
+				if err != nil {
+					return err
+				}
+
+				sp := model.Span{}
+				switch item.UserMeta() & 0x0F {
+				case jsonEncoding:
+					if err := json.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				case protoEncoding:
+					if err := proto.Unmarshal(val, &sp); err != nil {
+						return err
+					}
+				default:
+					return fmt.Errorf("Unknown encoding type: %04b", item.UserMeta()&0x0F)
+				}
+				spans = append(spans, &sp)
+			}
+			trace := &model.Trace{
+				Spans: spans,
+			}
+			traces = append(traces, trace)
+		}
+		return nil
+	})
+
+	// TODO Do the Unmarshal here so we can release the transaction earlier (we would pay with extra allocations..)
```

Comment:
is this still needed?

Reply:
Not really, it's not a bottleneck at the moment.